waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java

```diff
@@ -252,62 +255,90 @@ private boolean isWhitelisted(String databasePrefix, String database) {
+              for (TableMeta tableMeta : tables) {
+                if (isWhitelisted(mapping.getDatabasePrefix(), tableMeta.getDbName())) {
+                  combined.add(mapping.transformOutboundTableMeta(tableMeta));
+                }
               }
+            } catch (TException e) {
+              LOG.warn("Got exception fetching get_table_meta: {}", e.getMessage());
             }
-          } catch (TException e) {
-            LOG.warn("Got exception fetching get_table_meta: {}", e.getMessage());
-          }
+          };
+          executorService.submit(runnableTask);
         }
+        shutdownExecutorService(executorService);
         return combined;
       }
 
       @Override
       public List<String> getAllDatabases(String databasePattern) {
         List<String> combined = new ArrayList<>();
+        ExecutorService executorService = Executors.newCachedThreadPool();
+
         for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsByDbPattern(databasePattern)
             .entrySet()) {
-          try {
-            DatabaseMapping mapping = mappingWithPattern.getKey();
-            String pattern = mappingWithPattern.getValue();
-            List<String> databases = mapping.getClient().get_databases(pattern);
-            for (String database : databases) {
-              if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
-                combined.add(mapping.transformOutboundDatabaseName(database));
+          Runnable runnableTask = () -> {
+            try {
+              DatabaseMapping mapping = mappingWithPattern.getKey();
+              String pattern = mappingWithPattern.getValue();
+              List<String> databases = mapping.getClient().get_databases(pattern);
+              for (String database : databases) {
+                if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
+                  combined.add(mapping.transformOutboundDatabaseName(database));
+                }
               }
+            } catch (TException e) {
+              LOG.warn("Can't fetch databases by pattern: {}", e.getMessage());
             }
-          } catch (TException e) {
-            LOG.warn("Can't fetch databases by pattern: {}", e.getMessage());
-          }
+          };
+          executorService.submit(runnableTask);
         }
+        shutdownExecutorService(executorService);
         return combined;
       }
 
       @Override
       public List<String> getAllDatabases() {
         List<String> combined = new ArrayList<>();
-        for (DatabaseMapping mapping : databaseMappings()) {
-          try {
-            List<String> databases = mapping.getClient().get_all_databases();
-            for (String database : databases) {
-              if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
-                combined.add(mapping.transformOutboundDatabaseName(database));
+        ExecutorService executorService = Executors.newCachedThreadPool();
```

Comment:
We know exactly how many concurrent jobs we need so a fixed size thread pool is probably better.